server/verify.go

```diff
@@ -0,0 +1,80 @@
+package server
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/subtle"
+	"fmt"
+	"sort"
+
+	"github.com/google/go-tpm-tools/proto"
+	"github.com/google/go-tpm-tools/tpm2tools"
+	"github.com/google/go-tpm/tpm2"
+)
+
+// Verify will take in a public key (currently, only RSA is supported), a proto
+// (contains an attestation and a signature), and will check the attestation is
+// valid or not.
+// In addition to the basic RSA signature checking, this function will also check
+// the PCR digest for quoted data, which can insure that the given PCR state is the
+// same as when signing this attestation. This function will return the "Extra Data"
+// appended in the attestation when quoting if passed all verification.
+func Verify(pubkey crypto.PublicKey, quote *proto.Quote, pcrs *proto.Pcrs) ([]byte, error) {
+	sig, err := tpm2.DecodeSignature(bytes.NewBuffer(quote.GetRawSig()))
+	if err != nil {
+		return nil, fmt.Errorf("signature decoding failed: %v", err)
+	}
+	if rsaPubKey, ok := pubkey.(*rsa.PublicKey); ok {
+		switch sig.Alg {
+		case tpm2.AlgRSASSA:
+			hash, err := sig.RSA.HashAlg.Hash()
+			if err != nil {
+				return nil, err
+			}
+			hashConstructor := hash.New()
+			hashConstructor.Write(quote.GetQuote())
+			if err := rsa.VerifyPKCS1v15(rsaPubKey, hash, hashConstructor.Sum(nil), sig.RSA.Signature); err != nil {
+				return nil, fmt.Errorf("failed to verify the signature: %v", err)
+			}
+		default:
+			return nil, fmt.Errorf("given signature scheme is not supported: %x, only RSASSA-PKCS1-v1_5 is supported", sig.Alg)
+		}
+	} else {
```

Comment:
I would put the error branch earlier to make the code easier to read:
```go
rsaPubKey, ok := pubkey.(*rsa.PublicKey)
if !ok {
        // Sorry, only RSA works right now
}
if sig.Alg != tpm2.AlgRSASSA {
        // Sorry, only RSASSA works right now
}

//  Hashing and PCKS v1.5 verification

```

Reply:
done